ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2835,194 @@ public sealed class Bot : IAsyncDisposable {
+			// Level 5 is maximum level for card badges according to https://steamcommunity.com/tradingcards/faq
+			linkElements.AddRange(badgePage.SelectNodes("//div[contains(@class, 'badge_row') and .//div[@class='badge_info_description']/div[contains(text(), 'Level 5')]]/a[@class='badge_row_overlay']"));
+
+			if (linkElements.Count == 0) {
+				return new HashSet<uint>(0);
+			}
+
+			HashSet<uint> result = new HashSet<uint>(linkElements.Count);
+
+			foreach (string? badgeUri in linkElements.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					continue;
+				}
+
+				// URIs to foil badges are the same as for normal badges except they end with "?border=1"
+				string? appIDText = badgeUri.Replace("?border=1", "").Split('/', StringSplitOptions.RemoveEmptyEntries).Last();
+
+				if (!uint.TryParse(appIDText, out uint appID) || (appID == 0)) {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					continue;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private async Task SendCompletedSets() {
+			ISet<uint> appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+			if (appIDs.Count == 0) {
+				// Could not extract appIds for craftable badges - No need to load the inventory at this point
+				return;
+			}
+
+			HashSet<Steam.Asset> inventory;
+
+			try {
+				inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+					.Where(item => item.Tradable)
+					.Where(item => (item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard))
+					.Where(item => appIDs.Contains(item.RealAppID))
+					.ToHashSetAsync()
+					.ConfigureAwait(false);
+			} catch (HttpRequestException e) {
+				ArchiLogger.LogGenericWarningException(e);
+
+				return;
+			} catch (Exception e) {
+				ArchiLogger.LogGenericException(e);
+
+				return;
+			}
+
+			if (inventory.Count == 0) {
+				ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+				return;
+			}
+
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), uint> inventorySets = Trading.GetInventorySets(inventory).ToDictionary(entry => entry.Key, entry => entry.Value.FirstOrDefault());
+			HashSet<Steam.Asset> result = (await Utilities.InParallel(inventorySets.Select(async set => GetItemsForFullBadge(inventory, set.Key.RealAppID, set.Key.Type, set.Key.Rarity, await ArchiWebHandler.GetCardCountForGame(set.Key.RealAppID).ConfigureAwait(false), set.Value))).ConfigureAwait(false)).SelectMany(set => set).ToHashSet();
+
+			if (result.Count > 0) {
+				await Actions.SendInventory(result).ConfigureAwait(false);
+			}
+		}
+
+		internal static IEnumerable<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, uint appID, Steam.Asset.EType type, Steam.Asset.ERarity rarity, byte cardsPerSet, uint amountPerCard) {
```

Comment:
No point of this returning `IEnumerable`, we never throw away part of the results, just return a collection as usual - it's especially important as you allocate quite a lot of data below, data which can be discarded when the result is already calculated.

Reply:
Also add arguments checks which once again are missing.

Reply:
`IEnumerable` is due to usage `yield` keyword, but I can rewrite to manually collect in `HashSet`